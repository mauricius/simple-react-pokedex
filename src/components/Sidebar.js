import React from 'react'
import { Link } from 'react-router-dom'
import logo from '../pokeball.png'

class Sidebar extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      search: '',
    }
  }

  handleChange = (e) => {
    this.setState({
      search: e.target.value
    })
  }

  render() {
    const { show, changingPokemon, pokemon } = this.props

    const re = new RegExp(this.state.search, "i")

    const filteredPokemon = pokemon ? pokemon.filter((pokemon) => {
      return re.test(pokemon.name)
    }) : []

    return (
      <div className={(show ? 'flex' : 'hidden') + " z-50 fixed pin-y pin-l bg-grey-lighter w-4/5 md:w-full md:max-w-xs flex-none border-r-2 border-grey-light md:flex flex-col"}>
        <div className="mb-4">
          <div className="hidden md:flex md:justify-center">
            <Link to="/" className="flex justify-center">
              <img src={logo} alt="logo" />
            </Link>
          </div>
          <p className="text-center">
            <a href="https://github.com/mauricius" className="text-sm hover:text-grey-dark text-grey font-semibold">Visit on GitHub</a>
          </p>
        </div>
        <div className="p-8 flex-1 overflow-y-scroll scrolling-touch md:scrolling-auto">
          <nav className="text-base">
            <form className="mb-4">
              <div className="mb-4">
                <label className="block text-grey-dark uppercase text-sm font-bold mb-2" htmlFor="search">
                  Search Pokemon
                </label>
                <input className="shadow appearance-none border rounded w-full py-2 px-3 text-grey-darker leading-tight focus:outline-none focus:shadow-outline" type="search" placeholder="Search" value={this.state.search} onChange={this.handleChange} />
              </div>
            </form>
            <ul>
              {
                filteredPokemon ?
                filteredPokemon.map(pokemonster => (
                  <li className="mb-3" key={pokemonster.id}>
                    <Link to={'/pokemon/' + pokemonster.id} className="hover:cursor-pointer hover:text-red text-grey-darkest no-underline" onClick={changingPokemon}>
                      { pokemonster.name }
                    </Link>
                  </li>
                ))
                : (
                  <li>Loading...</li>
                )
              }
            </ul>
          </nav>
        </div>
      </div>
    )
  }
}

export default Sidebar