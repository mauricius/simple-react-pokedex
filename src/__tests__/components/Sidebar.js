import React from 'react'
import { MemoryRouter as Router, withRouter } from 'react-router-dom'
import Sidebar from '../../components/Sidebar'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

describe('test Sidebar component', function () {

  it('renders correctly without pokemon', () => {
    const component = renderer
      .create(<Router><Sidebar show={true} /></Router>)
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it('renders correctly with a list of pokemon', () => {
    const pokemon = [
      {
        id: 1,
        name: 'bulbasaur'
      }, {
        id: 2,
        name: 'ivysaur'
      }
    ]

    const component = renderer
      .create(<Router><Sidebar show={true} pokemon={pokemon} /></Router>)
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it('renders an empty <Sidebar /> components', () => {
    const wrapper = shallow(<Sidebar show={true} pokemon={[]} />);
    expect(wrapper.contains(<ul></ul>)).toBeTruthy
  })

  it('filters Pokemon on <Sidebar /> component if a query is typed in the search input', () => {
    const pokemon = [
      {
        id: 1,
        name: 'bulbasaur'
      }, {
        id: 2,
        name: 'ivysaur'
      }
    ]

    const wrapper = shallow(<Sidebar show={true} pokemon={pokemon} />);

    expect(wrapper.find('.mb-3').length).toEqual(2)
    wrapper.find('[type="search"]').simulate('change', { target: { value: 'bulba' }})
    expect(wrapper.find('.mb-3').length).toEqual(1)
  })
})
