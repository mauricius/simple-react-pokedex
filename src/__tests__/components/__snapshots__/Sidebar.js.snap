// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Sidebar component renders correctly with a list of pokemon 1`] = `
<div
  className="flex z-50 fixed pin-y pin-l bg-grey-lighter w-4/5 md:w-full md:max-w-xs flex-none border-r-2 border-grey-light md:flex flex-col"
>
  <div
    className="mb-4"
  >
    <div
      className="hidden md:flex md:justify-center"
    >
      <a
        className="flex justify-center"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="logo"
          src="pokeball.png"
        />
      </a>
    </div>
    <p
      className="text-center"
    >
      <a
        className="text-sm hover:text-grey-dark text-grey font-semibold"
        href="https://github.com/mauricius"
      >
        Visit on GitHub
      </a>
    </p>
  </div>
  <div
    className="p-8 flex-1 overflow-y-scroll scrolling-touch md:scrolling-auto"
  >
    <nav
      className="text-base"
    >
      <form
        className="mb-4"
      >
        <div
          className="mb-4"
        >
          <label
            className="block text-grey-dark uppercase text-sm font-bold mb-2"
            htmlFor="search"
          >
            Search Pokemon
          </label>
          <input
            className="shadow appearance-none border rounded w-full py-2 px-3 text-grey-darker leading-tight focus:outline-none focus:shadow-outline"
            onChange={[Function]}
            placeholder="Search"
            type="search"
            value=""
          />
        </div>
      </form>
      <ul>
        <li
          className="mb-3"
        >
          <a
            className="hover:cursor-pointer hover:text-red text-grey-darkest no-underline"
            href="/pokemon/1"
            onClick={[Function]}
          >
            bulbasaur
          </a>
        </li>
        <li
          className="mb-3"
        >
          <a
            className="hover:cursor-pointer hover:text-red text-grey-darkest no-underline"
            href="/pokemon/2"
            onClick={[Function]}
          >
            ivysaur
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`test Sidebar component renders correctly without pokemon 1`] = `
<div
  className="flex z-50 fixed pin-y pin-l bg-grey-lighter w-4/5 md:w-full md:max-w-xs flex-none border-r-2 border-grey-light md:flex flex-col"
>
  <div
    className="mb-4"
  >
    <div
      className="hidden md:flex md:justify-center"
    >
      <a
        className="flex justify-center"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="logo"
          src="pokeball.png"
        />
      </a>
    </div>
    <p
      className="text-center"
    >
      <a
        className="text-sm hover:text-grey-dark text-grey font-semibold"
        href="https://github.com/mauricius"
      >
        Visit on GitHub
      </a>
    </p>
  </div>
  <div
    className="p-8 flex-1 overflow-y-scroll scrolling-touch md:scrolling-auto"
  >
    <nav
      className="text-base"
    >
      <form
        className="mb-4"
      >
        <div
          className="mb-4"
        >
          <label
            className="block text-grey-dark uppercase text-sm font-bold mb-2"
            htmlFor="search"
          >
            Search Pokemon
          </label>
          <input
            className="shadow appearance-none border rounded w-full py-2 px-3 text-grey-darker leading-tight focus:outline-none focus:shadow-outline"
            onChange={[Function]}
            placeholder="Search"
            type="search"
            value=""
          />
        </div>
      </form>
      <ul />
    </nav>
  </div>
</div>
`;
